File: scripts/greek-carnivore-content-generator.py
#!/usr/bin/env python3
"""
Enhanced Greek Carnivore Content Generator
Creates contextual descriptions based on video content and messaging
"""
import json
import random
from datetime import datetime
from typing import Dict, List, Optional

class GreekCarnivoreContentGenerator:
    def __init__(self):
        # Greek carnivore-specific content templates
        self.templates = {
            "diet_testing": {
                "hooks": [
                    "Δοκιμάζω την κάρνιβορ διατροφή για 30 μέρες",
                    "Η κάρνιβορ διατροφή άλλαξε τη ζωή μου",
                    "Τι συμβαίνει όταν τρως ΜΟΝΟ κρέας;",
                    "30 μέρες μόνο κρέας - Τα αποτελέσματα θα σε σοκάρουν"
                ],
                "descriptions": [
                    """🥩 Δοκιμάζω την κάρνιβορ διατροφή! 
Μόνο κρέας για έναν μήνα - δες τι συμβαίνει στο σώμα μου.
Περιγραφή περιήγησης για την κάρνιβορ ζωή! 💪

#ΚάρνιβορΔιατροφή #ΕλληνικήΚάρνιβορ #ΜόνοΚρέας #ΥγιεινήΖωή""",
                    
                    """🔥 Η κάρνιβορ διατροφή σε δράση!
Ξεκίνησα να αισθάνεσαι περιήγηση για το σώμα σου με εύκολες και πρακτικές κινήσεις.
Είσαι έτοιμος; 

💪 #ΑγαπήστεΤοΣώματαΣας #ΑπώλειαΒάρους #ΓρεεκΚάρνιβορ""",
                    
                    """🥩 ΚΡΕΑΣ = ΔΥΝΑΜΗ
Δες γιατί η κάρνιβορ διατροφή είναι το μυστικό για:
✅ Απώλεια βάρους
✅ Περισσότερη ενέργεια  
✅ Καλύτερη υγεία
✅ Ψυχική διαύγεια

#ΚάρνιβορΣτυλ #ΕλλάδαΚάρνιβορ #ΥγιεινήΔιατροφή"""
                ]
            },
            
            "transformation": {
                "hooks": [
                    "Πώς έχασα 15 κιλά σε 3 μήνες",
                    "Η μεταμόρφωσή μου με την κάρνιβορ",
                    "Από 90 κιλά σε 75 - Η ιστορία μου",
                    "Αυτό που κάνει η κάρνιβορ στο σώμα"
                ],
                "descriptions": [
                    """🔥 ΜΕΤΑΜΟΡΦΩΣΗ ALERT! 
Από 90 κιλά σε 75 σε 3 μήνες - ΜΟΝΟ με κάρνιβορ διατροφή!

✨ Τι άλλαξε:
• Απώλεια βάρους χωρίς πείνα
• Ενέργεια στα ύψη
• Καθαρό δέρμα
• Καλύτερος ύπνος

#ΚάρνιβορΜεταμόρφωση #ΑπώλειαΒάρους #ΕλληνικήΔιατροφή""",
                    
                    """💪 Η ΑΛΗΤΕΙΑ για την κάρνιβορ διατροφή!
Δες τη δική μου μεταμόρφωση και μάθε:
🥩 Τι τρώω κάθε μέρα
⏰ Πότε τρώω  
💡 Γιατί λειτουργεί

Είσαι έτοιμος για αλλαγή;

#ΚάρνιβορΖωή #ΕλλάδαΚάρνιβορ #ΥγιεινήΖωή"""
                ]
            },
            
            "daily_tips": {
                "hooks": [
                    "5 λόγοι να ξεκινήσεις κάρνιβορ ΣΗΜΕΡΑ",
                    "Το μυστικό που δεν σου λένε οι γιατροί",
                    "Γιατί το κρέας είναι η καλύτερη τροφή",
                    "Αυτό που δε ξέρεις για τη ζωική πρωτεΐνη"
                ],
                "descriptions": [
                    """🥩 ΚΑΡΝΙΒΟΡ ΣΥΜΒΟΥΛΗ ΤΗΣ ΗΜΕΡΑΣ!

Σήμερα μαθαίνουμε γιατί το κρέας είναι το ΤΕΛΕΙΟ φαγητό:
✅ Πλήρης πρωτεΐνη
✅ Βιταμίνες Β12 & σίδηρος
✅ Μηδέν υδατάνθρακες
✅ Φυσική κετόζη

Δοκίμασες ήδη; 

#ΚάρνιβορΣυμβουλές #ΕλληνικήΚάρνιβορ #ΥγιεινήΔιατροφή""",
                    
                    """💡 ΤΟ ΞΕΡΕΣ ΟΤΙ...
Η κάρνιβορ διατροφή είναι η πιο αρχαία διατροφή του ανθρώπου;

Οι πρόγονοί μας έτρωγαν ΜΟΝΟ ζωικές τροφές για χιλιάδες χρόνια!
Γι' αυτό το σώμα μας τα "θυμάται" αυτά τα φαγητά.

#ΚάρνιβορΙστορία #ΑρχαίαΔιατροφή #ΕλλάδαΚάρνιβορ"""
                ]
            },
            
            "food_showcase": {
                "hooks": [
                    "Το ΤΕΛΕΙΟ κάρνιβορ γεύμα",
                    "Τι τρώω σε μία μέρα - Κάρνιβορ edition",
                    "Ribeye vs Μοσχάρι - Ποιο είναι καλύτερο;",
                    "Η αλήθεια για τα όργανα (liver, heart)"
                ],
                "descriptions": [
                    """🍖 ΚΑΡΝΙΒΟΡ MEAL PREP!

Δες τι τρώω σε μία τυπική κάρνιβορ μέρα:
🌅 Πρωί: Αυγά + μπέικον
☀️ Μεσημέρι: Ribeye steak  
🌙 Βράδυ: Λάμπ τσοπς

Απλό, νόστιμο, αποτελεσματικό!

#ΚάρνιβορΦαγητό #ΜealPrep #ΕλληνικήΚάρνιβορ #Steak""",
                    
                    """🥩 RIBEYE = Ο ΒΑΣΙΛΙΑΣ των κρεάτων!

Γιατί το ribeye είναι το #1 κάρνιβορ φαγητό:
🔥 Υψηλά λιπαρά για ενέργεια
💪 Πλούσιο σε πρωτεΐνη  
😋 Απίστευτη γεύση
🧠 Ωμέγα-3 για τον εγκέφαλο

Έχεις δοκιμάσει;

#Ribeye #ΚάρνιβορΣτυλ #ΕλληνικόΚρέας"""
                ]
            }
        }
        
        self.hashtags_greek = [
            "#ΚάρνιβορΔιατροφή", "#ΕλληνικήΚάρνιβορ", "#ΜόνοΚρέας", 
            "#ΑπώλειαΒάρους", "#ΥγιεινήΖωή", "#ΚάρνιβορΣτυλ",
            "#ΕλλάδαΚάρνιβορ", "#ΚρεατοΦάγος", "#ΦυσικήΔιατροφή",
            "#ΠαλαιολιθικήΔιατροφή", "#ΚάρνιβορΜεταμόρφωση"
        ]
        
        self.english_hashtags = [
            "#Carnivore", "#CarnivoreLife", "#MeatOnly", "#CarnivoreDiet",
            "#ZeroCarb", "#AnimalBased", "#CarnivoreResults", "#GreekCarnivore",
            "#CarnivoreTransformation", "#MeatHeals"
        ]

    def generate_reel_description(self, 
                                 video_context: str, 
                                 content_type: str = "diet_testing",
                                 include_english: bool = True) -> Dict:
        """
        Generate Greek carnivore reel description based on video context
        
        Args:
            video_context: What's happening in the video (Greek or English)
            content_type: Type of content (diet_testing, transformation, daily_tips, food_showcase)
            include_english: Whether to include English hashtags
        """
        
        if content_type not in self.templates:
            content_type = "diet_testing"  # Default fallback
        
        template = self.templates[content_type]
        
        # Select hook and description
        hook = random.choice(template["hooks"])
        description = random.choice(template["descriptions"])
        
        # Add contextual elements based on video_context
        contextual_additions = self._get_contextual_additions(video_context, content_type)
        
        # Combine hashtags
        hashtags = random.sample(self.hashtags_greek, 4)  # 4 Greek hashtags
        if include_english:
            hashtags.extend(random.sample(self.english_hashtags, 3))  # 3 English hashtags
        
        final_description = description
        if contextual_additions:
            final_description += f"\n\n{contextual_additions}"
        
        final_description += f"\n\n{' '.join(hashtags)}"
        
        return {
            "hook": hook,
            "description": final_description,
            "hashtags": hashtags,
            "content_type": content_type,
            "character_count": len(final_description),
            "estimated_performance": self._estimate_performance(content_type, final_description)
        }

    def _get_contextual_additions(self, video_context: str, content_type: str) -> str:
        """Add context-specific content based on what's in the video"""
        
        context_lower = video_context.lower()
        additions = []
        
        # Check for specific elements in the video
        if "δοκιμάζει" in context_lower or "testing" in context_lower:
            additions.append("🎯 Παρακολούθησε το ταξίδι μου!")
            
        if "κρέας" in context_lower or "meat" in context_lower:
            additions.append("🥩 Quality κρέας = Quality αποτελέσματα")
            
        if "μεταμόρφωση" in context_lower or "transformation" in context_lower:
            additions.append("📸 Πριν/μετά φωτογραφίες σύντομα!")
            
        if "ribeye" in context_lower or "steak" in context_lower:
            additions.append("🔥 Grass-fed όταν γίνεται!")
            
        if "αποτελέσματα" in context_lower or "results" in context_lower:
            additions.append("📊 Τα νούμερα δε λένε ψέματα!")
        
        return " ".join(additions) if additions else ""

    def _estimate_performance(self, content_type: str, description: str) -> float:
        """Estimate performance score based on content analysis"""
        
        base_scores = {
            "transformation": 9.0,
            "diet_testing": 8.5,
            "food_showcase": 8.0,
            "daily_tips": 7.5
        }
        
        score = base_scores.get(content_type, 7.0)
        
        # Boost for engagement elements
        if "?" in description:
            score += 0.3
        if any(emoji in description for emoji in ["🥩", "🔥", "💪", "✅"]):
            score += 0.2
        if len(description) > 150 and len(description) < 300:  # Optimal length
            score += 0.3
        if "#" in description:
            score += 0.2
            
        return min(score, 10.0)

    def suggest_improvements(self, current_description: str) -> List[str]:
        """Suggest improvements for existing descriptions"""
        
        suggestions = []
        
        if len(current_description) > 400:
            suggestions.append("Μείωσε το μήκος - κάτω από 400 χαρακτήρες είναι καλύτερο")
            
        if not any(emoji in current_description for emoji in ["🥩", "🔥", "💪", "✅"]):
            suggestions.append("Πρόσθεσε emojis για περισσότερη προσοχή")
            
        if "#" not in current_description:
            suggestions.append("Πρόσθεσε hashtags για καλύτερη προβολή")
            
        if "?" not in current_description:
            suggestions.append("Πρόσθεσε ερώτηση για engagement")
            
        if not any(word in current_description.lower() for word in ["κάρνιβορ", "carnivore", "κρέας"]):
            suggestions.append("Συμπέριλαβε λέξεις-κλειδιά: κάρνιβορ, κρέας")
            
        return suggestions

# Example usage for the reel in the image
def generate_for_image_content():
    """Generate content for the specific reel shown in the image"""
    
    generator = GreekCarnivoreContentGenerator()
    
    # Based on the image: "Δοκιμάζει την Κάρνιβορ διατροφή"
    video_context = "Δοκιμάζει την κάρνιβορ διατροφή - testing carnivore diet lifestyle transformation"
    
    result = generator.generate_reel_description(
        video_context=video_context,
        content_type="diet_testing",
        include_english=True
    )
    
    return result

if __name__ == "__main__":
    # Test with the image content
    result = generate_for_image_content()
    print("GENERATED GREEK CARNIVORE REEL DESCRIPTION:")
    print("=" * 50)
    print(f"Hook: {result['hook']}")
    print(f"\nDescription:\n{result['description']}")
    print(f"\nCharacter Count: {result['character_count']}")
    print(f"Estimated Performance: {result['estimated_performance']}/10")

File: src/lib/greekCarnivoreContentGenerator.ts
// Ported from scripts/greek-carnivore-content-generator.py

export type ContentType = 'diet_testing' | 'transformation' | 'daily_tips' | 'food_showcase';

export interface GeneratedContent {
  hook: string;
  description: string;
  hashtags: string[];
  content_type: ContentType;
  character_count: number;
  estimated_performance: number;
}

const TEMPLATES: Record<ContentType, { hooks: string[]; descriptions: string[] }> = {
  diet_testing: {
    hooks: [
      'Δοκιμάζω την κάρνιβορ διατροφή για 30 μέρες',
      'Η κάρνιβορ διατροφή άλλαξε τη ζωή μου',
      'Τι συμβαίνει όταν τρως ΜΟΝΟ κρέας;',
      '30 μέρες μόνο κρέας - Τα αποτελέσματα θα σε σοκάρουν',
    ],
    descriptions: [
      '🥩 Δοκιμάζω την κάρνιβορ διατροφή!\nΜόνο κρέας για έναν μήνα - δες τι συμβαίνει στο σώμα μου.\nΠεριγραφή περιήγησης για την κάρνιβορ ζωή! 💪\n\n#ΚάρνιβορΔιατροφή #ΕλληνικήΚάρνιβορ #ΜόνοΚρέας #ΥγιεινήΖωή',
      '🔥 Η κάρνιβορ διατροφή σε δράση!\nΞεκίνησα να αισθάνεσαι περιήγηση για το σώμα σου με εύκολες και πρακτικές κινήσεις.\nΕίσαι έτοιμος;\n\n💪 #ΑγαπήστεΤοΣώματαΣας #ΑπώλειαΒάρους #ΓρεεκΚάρνιβορ',
      '🥩 ΚΡΕΑΣ = ΔΥΝΑΜΗ\nΔες γιατί η κάρνιβορ διατροφή είναι το μυστικό για:\n✅ Απώλεια βάρους\n✅ Περισσότερη ενέργεια\n✅ Καλύτερη υγεία\n✅ Ψυχική διαύγεια\n\n#ΚάρνιβορΣτυλ #ΕλλάδαΚάρνιβορ #ΥγιεινήΔιατροφή',
    ],
  },
  transformation: {
    hooks: [
      'Πώς έχασα 15 κιλά σε 3 μήνες',
      'Η μεταμόρφωσή μου με την κάρνιβορ',
      'Από 90 κιλά σε 75 - Η ιστορία μου',
      'Αυτό που κάνει η κάρνιβορ στο σώμα',
    ],
    descriptions: [
      '🔥 ΜΕΤΑΜΟΡΦΩΣΗ ALERT!\nΑπό 90 κιλά σε 75 σε 3 μήνες - ΜΟΝΟ με κάρνιβορ διατροφή!\n\n✨ Τι άλλαξε:\n• Απώλεια βάρους χωρίς πείνα\n• Ενέργεια στα ύψη\n• Καθαρό δέρμα\n• Καλύτερος ύπνος\n\n#ΚάρνιβορΜεταμόρφωση #ΑπώλειαΒάρους #ΕλληνικήΔιατροφή',
      '💪 Η ΑΛΗΤΕΙΑ για την κάρνιβορ διατροφή!\nΔες τη δική μου μεταμόρφωση και μάθε:\n🥩 Τι τρώω κάθε μέρα\n⏰ Πότε τρώω\n💡 Γιατί λειτουργεί\n\nΕίσαι έτοιμος για αλλαγή;\n\n#ΚάρνιβορΖωή #ΕλλάδαΚάρνιβορ #ΥγιεινήΖωή',
    ],
  },
  daily_tips: {
    hooks: [
      '5 λόγοι να ξεκινήσεις κάρνιβορ ΣΗΜΕΡΑ',
      'Το μυστικό που δεν σου λένε οι γιατροί',
      'Γιατί το κρέας είναι η καλύτερη τροφή',
      'Αυτό που δε ξέρεις για τη ζωική πρωτεΐνη',
    ],
    descriptions: [
      '🥩 ΚΑΡΝΙΒΟΡ ΣΥΜΒΟΥΛΗ ΤΗΣ ΗΜΕΡΑΣ!\n\nΣήμερα μαθαίνουμε γιατί το κρέας είναι το ΤΕΛΕΙΟ φαγητό:\n✅ Πλήρης πρωτεΐνη\n✅ Βιταμίνες Β12 & σίδηρος\n✅ Μηδέν υδατάνθρακες\n✅ Φυσική κετόζη\n\nΔοκίμασες ήδη;\n\n#ΚάρνιβορΣυμβουλές #ΕλληνικήΚάρνιβορ #ΥγιεινήΔιατροφή',
      '💡 ΤΟ ΞΕΡΕΣ ΟΤΙ...\nΗ κάρνιβορ διατροφή είναι η πιο αρχαία διατροφή του ανθρώπου;\n\nΟι πρόγονοί μας έτρωγαν ΜΟΝΟ ζωικές τροφές για χιλιάδες χρόνια!\nΓι\' αυτό το σώμα μας τα "θυμάται" αυτά τα φαγητά.\n\n#ΚάρνιβορΙστορία #ΑρχαίαΔιατροφή #ΕλλάδαΚάρνιβορ',
    ],
  },
  food_showcase: {
    hooks: [
      'Το ΤΕΛΕΙΟ κάρνιβορ γεύμα',
      'Τι τρώω σε μία μέρα - Κάρνιβορ edition',
      'Ribeye vs Μοσχάρι - Ποιο είναι καλύτερο;',
      'Η αλήθεια για τα όργανα (liver, heart)',
    ],
    descriptions: [
      '🍖 ΚΑΡΝΙΒΟΡ MEAL PREP!\n\nΔες τι τρώω σε μία τυπική κάρνιβορ μέρα:\n🌅 Πρωί: Αυγά + μπέικον\n☀️ Μεσημέρι: Ribeye steak\n🌙 Βράδυ: Λάμπ τσοπς\n\nΑπλό, νόστιμο, αποτελεσματικό!\n\n#ΚάρνιβορΦαγητό #MealPrep #ΕλληνικήΚάρνιβορ #Steak',
      '🥩 RIBEYE = Ο ΒΑΣΙΛΙΑΣ των κρεάτων!\n\nΓιατί το ribeye είναι το #1 κάρνιβορ φαγητό:\n🔥 Υψηλά λιπαρά για ενέργεια\n💪 Πλούσιο σε πρωτεΐνη\n😋 Απίστευτη γεύση\n🧠 Ωμέγα-3 για τον εγκέφαλο\n\nΈχεις δοκιμάσει;\n\n#Ribeye #ΚάρνιβορΣτυλ #ΕλληνικόΚρέας',
    ],
  },
};

const GREEK_HASHTAGS = [
  '#ΚάρνιβορΔιατροφή', '#ΕλληνικήΚάρνιβορ', '#ΜόνοΚρέας',
  '#ΑπώλειαΒάρους', '#ΥγιεινήΖωή', '#ΚάρνιβορΣτυλ',
  '#ΕλλάδαΚάρνιβορ', '#ΚρεατοΦάγος', '#ΦυσικήΔιατροφή',
  '#ΠαλαιολιθικήΔιατροφή', '#ΚάρνιβορΜεταμόρφωση',
];

const ENGLISH_HASHTAGS = [
  '#Carnivore', '#CarnivoreLife', '#MeatOnly', '#CarnivoreDiet',
  '#ZeroCarb', '#AnimalBased', '#CarnivoreResults', '#GreekCarnivore',
  '#CarnivoreTransformation', '#MeatHeals',
];

function pick<T>(arr: T[]): T {
  return arr[Math.floor(Math.random() * arr.length)];
}

function sample<T>(arr: T[], n: number): T[] {
  return [...arr].sort(() => 0.5 - Math.random()).slice(0, n);
}

function getContextualAdditions(videoContext: string): string {
  const ctx = videoContext.toLowerCase();
  const additions: string[] = [];
  if (ctx.includes('δοκιμάζει') || ctx.includes('testing')) additions.push('🎯 Παρακολούθησε το ταξίδι μου!');
  if (ctx.includes('κρέας') || ctx.includes('meat')) additions.push('🥩 Quality κρέας = Quality αποτελέσματα');
  if (ctx.includes('μεταμόρφωση') || ctx.includes('transformation')) additions.push('📸 Πριν/μετά φωτογραφίες σύντομα!');
  if (ctx.includes('ribeye') || ctx.includes('steak')) additions.push('🔥 Grass-fed όταν γίνεται!');
  if (ctx.includes('αποτελέσματα') || ctx.includes('results')) additions.push('📊 Τα νούμερα δε λένε ψέματα!');
  return additions.join(' ');
}

function estimatePerformance(contentType: ContentType, description: string): number {
  const base: Record<ContentType, number> = {
    transformation: 9.0,
    diet_testing: 8.5,
    food_showcase: 8.0,
    daily_tips: 7.5,
  };
  let score = base[contentType];
  if (description.includes('?')) score += 0.3;
  if (['🥩', '🔥', '💪', '✅'].some(e => description.includes(e))) score += 0.2;
  if (description.length > 150 && description.length < 300) score += 0.3;
  if (description.includes('#')) score += 0.2;
  return Math.min(score, 10.0);
}

export function generateReelDescription(
  videoContext: string,
  contentType: ContentType = 'diet_testing',
  includeEnglish = true
): GeneratedContent {
  const template = TEMPLATES[contentType];
  const hook = pick(template.hooks);
  let description = pick(template.descriptions);

  const contextual = getContextualAdditions(videoContext);
  if (contextual) description += `\n\n${contextual}`;

  const hashtags = [
    ...sample(GREEK_HASHTAGS, 4),
    ...(includeEnglish ? sample(ENGLISH_HASHTAGS, 3) : []),
  ];
  description += `\n\n${hashtags.join(' ')}`;

  return {
    hook,
    description,
    hashtags,
    content_type: contentType,
    character_count: description.length,
    estimated_performance: estimatePerformance(contentType, description),
  };
}

export function suggestImprovements(description: string): string[] {
  const suggestions: string[] = [];
  if (description.length > 400) suggestions.push('Μείωσε το μήκος - κάτω από 400 χαρακτήρες είναι καλύτερο');
  if (!['🥩', '🔥', '💪', '✅'].some(e => description.includes(e))) suggestions.push('Πρόσθεσε emojis για περισσότερη προσοχή');
  if (!description.includes('#')) suggestions.push('Πρόσθεσε hashtags για καλύτερη προβολή');
  if (!description.includes('?')) suggestions.push('Πρόσθεσε ερώτηση για engagement');
  if (!['κάρνιβορ', 'carnivore', 'κρέας'].some(w => description.toLowerCase().includes(w))) suggestions.push('Συμπέριλαβε λέξεις-κλειδιά: κάρνιβορ, κρέας');
  return suggestions;
}

export const CONTENT_TYPES: { value: ContentType; label: string }[] = [
  { value: 'diet_testing', label: 'Diet Testing' },
  { value: 'transformation', label: 'Transformation' },
  { value: 'daily_tips', label: 'Daily Tips' },
  { value: 'food_showcase', label: 'Food Showcase' },
];
